lib/ansible/modules/files/ini_file.py

```diff
@@ -160,77 +166,140 @@ def do_ini(module, filename, section=None, option=None, value=None,
         ini_lines[-1] += '\n'
         changed = True
 
-    # append a fake section line to simplify the logic
-    ini_lines.append('[')
+    if not section or section == 'null':
+        # Managing options within no section
+        # Regex to look for a section
+        r_section_string = '^\s*\['
+        r_section = re.compile(r_section_string)
 
-    within_section = not section
-    section_start = 0
-    msg = 'OK'
-    if no_extra_spaces:
-        assignment_format = '%s=%s\n'
-    else:
-        assignment_format = '%s = %s\n'
+        # Regex to look for the option key
+        r_key_string = '^\s*%s\s*=\s*' % option
+        r_key = re.compile(r_key_string)
 
-    for index, line in enumerate(ini_lines):
-        if line.startswith('[%s]' % section):
-            within_section = True
-            section_start = index
-        elif line.startswith('['):
-            if within_section:
-                if state == 'present':
-                    # insert missing option line at the end of the section
-                    for i in range(index, 0, -1):
-                        # search backwards for previous non-blank or non-comment line
-                        if not re.match(r'^[ \t]*([#;].*)?$', ini_lines[i - 1]):
-                            ini_lines.insert(i, assignment_format % (option, value))
-                            msg = 'option added'
-                            changed = True
-                            break
-                elif state == 'absent' and not option:
-                    # remove the entire section
-                    del ini_lines[section_start:index]
-                    msg = 'section removed'
-                    changed = True
-                break
+        # Look for the first section line
+        section_found = False
+
+        # Looking for every section defined
+        sections_start_line = [index for index, line in enumerate(ini_lines) if r_section.match(line)]
+
+        try:
+            # line of the first section...
+            first_section_line = sections_start_line[0]
+            section_found = True
+        except IndexError:
+            # ...or, if no section found, the scope to look for the option is the whole file
+            first_section_line = len(ini_lines)
+
+        # Looking for the option
+        option_found = False
+        if first_section_line > 0:
```

Comment:
This condition doesn't seem required: `[:0]` returns an empty list, code below won't break when `first_section_line` is equal to `0`.